RevenueCatUI/Purchasing/PurchaseHandler.swift

```diff
@@ -96,13 +106,62 @@ extension PurchaseHandler {
         return result
     }
 
+    func trackPaywallView(_ eventData: PaywallEvent.Data) {
+        self.eventData = eventData
+        self.trackEvent(PaywallEvent.view)
+    }
+
+    func trackPaywallClose() {
+        self.trackEvent(PaywallEvent.close)
+    }
+
+    fileprivate func trackCancelledPurchase() {
+        self.trackEvent(PaywallEvent.cancel)
+    }
+
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, *)
+extension PurchaseHandler {
+
     /// Creates a copy of this `PurchaseHandler` wrapping the purchase and restore blocks.
     func map(
         purchase: @escaping (@escaping PurchaseBlock) -> PurchaseBlock,
         restore: @escaping (@escaping RestoreBlock) -> RestoreBlock
     ) -> Self {
         return .init(purchase: purchase(self.purchaseBlock),
-                     restorePurchases: restore(self.restoreBlock))
+                     restorePurchases: restore(self.restoreBlock),
+                     trackEvent: self.trackEventBlock)
+    }
+
+    func map(
+        trackEvent: @escaping (@escaping TrackEventBlock) -> TrackEventBlock
+    ) -> Self {
+        return .init(
+            purchase: self.purchaseBlock,
+            restorePurchases: self.restoreBlock,
+            trackEvent: trackEvent(self.trackEventBlock)
+        )
+    }
+
+}
+
+// MARK: - Private
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, *)
+private extension PurchaseHandler {
+
+    func trackEvent(_ eventCreator: (PaywallEvent.Data) -> PaywallEvent) {
+        guard let data = self.eventData else {
+            Logger.warning(Strings.attempted_to_track_event_with_missing_data)
```

Comment:
This is covered by tests. I tried different approaches of where to create the session ID to avoid this, but this was the simplest.

Reply:
Could you elaborate on what problem this approach solves? I'm trying to understand how we ended up with trackEvent taking a function that turns data into an event, and I'm sure there's a reasonable explanation but Fried Brain Fridays

Reply:
This method relies on `self.eventData`, because that needs to be saved throughout the lifetime of a `PaywallView` (since it has the session ID, dark mode, locale, etc).
The closure parameter is just a way to avoid duplicated code. It takes `self.eventData` (which is a `PaywallEvent.Data`) and wraps it into a `PaywallEvent`.
So it gets used like this:
```
self.trackEvent(PaywallEvent.view)
self.trackEvent(PaywallEvent.cancel)
self.trackEvent(PaywallEvent.close)
```

Reply:
it took me a few mins to understand the implementation tbh. I got confused with the closure, the fact that there are a few things called trackEvent and the mapping on the init. 

Would this work? 

- always pass in the full paywall data for events, including cancel and close. The event reporting layer has that data after all, and that protects against unexpected cases where cancel comes before view or something (we're delegating into background priority tasks, and although I imagine things happen in order, in case of a freak accident maybe it'd be safer to always have the data). 
- remove the closures, making the implementation of this method something like
```swift
    func track(event: PaywallEvent) {
        Task.detached(priority: .background) { [block = self.trackEventBlock] in
            await purchases.track(paywallEvent: event)
        }
    }
```

and called like 
```swift
func trackPaywallView(_ eventData: PaywallEvent.Data) {
    self.track(event: PaywallEvent.view(eventData))
}
...
self.purchaseHandler.trackPaywallView(self.eventData)
```

Reply:
we'd need to also pass in the paywall data when calling purchase and restore, but I think that'd be fine? the caller knows the paywall information anyway

Reply:
Refactored